### PR TITLE
feat(core): receive query result using async iterable

### LINKED ISSUE #581: Ability to query rows using async iterables
__Proposal:__
A `QueryApi#iterateRows()` method which allows you to [asynchronously iterate](https://exploringjs.com/impatient-js/ch_async-iteration.html) over rows using built-in JavaScript language features, and to automatically trigger cancellation, backpressure, and error handling based on my program's behavior, instead of having to explicitly opt into it.

__Current behavior:__
The `QueryApi#queryRows()` method has a few drawbacks. Callbacks unnecessarily compartmentalize your logic and take you out of the flow of your program, including any try/catch blocks. Plus, there's no obvious way to do backpressure. Finally, I discovered `useCancellable` by perusing this lib's `.d.ts` file, but I'm not sure from reading docs how to use it.

```ts
queryApi.queryRows(query, {
  next: (row, tableMeta) => {
    // Do something with `row` and `tableMeta`...
  },
  error: (error) => {
    // Handle errors here
  },
  complete: () => {
    // Done
  },
  useCancellable: (cancelable) => {
    // How does this work? Maybe this?
    if (condition) {
      cancelable.cancel();
    }
  },
});
```

__Desired behavior:__
I'd like to be able to do this.

```ts
for await (const { row, tableMeta } of queryApi.iterateRows(query)) {
  // Do something with `row` and `tableMeta`...
  // Any `break`, `throw`, or `return` in here triggers cancellation
  // Any `throw` in here propagates to nearest try/catch
  // Any `await` in here creates backpressure
}
// Done
```

This lib can universally implement cancellation, backpressure, and error handling on behalf of users thusly:

- **Cancellation:** Implement or override the returned async iterable's `.return()` method to perform any cleanup or abort.
- **Backpressure:** Depends on how this lib consumes source data. If consuming a Node.js source stream for example, it could throttle it based on whether the client has finished processing the latest row.
- **Error handling:** Simply return a rejected promise from `iterator.next()`.

__Alternatives considered:__
Can continue using the existing callback API.

__Use case:__
Allows me to consume rows using idiomatic, built-in JavaScript language features, and to compose this lib more naturally with other promise-based and async-iterator-based utilities, and to avoid eating memory if my code can't consume data as fast as this lib produces it.

### PR DESCRIPTION
**Pull request by sranka**
Fixes #581

The main goal of this PR is to allow developers to consume query results using async iterable:

```js
for await (const {values, tableMeta} of queryApi.iterateRows(fluxQuery)){
   // consume a particular row at will 
  console.log(JSON.stringify(tableMeta.toObject(values)))
}
```
All the details are demonstrated in updated examples. The present approach (with an observable) is still valid in many use cases, it is however more verbose and a bit harder to use in an asynchronous context. 

## Proposed Changes

- allow receiving query results using async iterable (`for-await` loop)
   - enhance transport interface with an 'iterate' member function that returns async iterable of UInt8Array buffers,
     implement it in both node.js and browser transport
     - the latest node v14 is required because of the maturity of the response iterable implementation
   - expose the possibility to use AbortSignal in both node and browser transport (for the sake of completeness)
   - enhance `AnnotatedCSVResponse` to return iterable of lines or rows
   - enhance `QueryApi` to return CSV results lines or rows using async iterable (new `iterateLines` and `iterateRows` method)
- change examples to demonstrate usage of async iterable
   - examples recommend using node v16
   - `query.ts` example was reorganized a bit so that top-level await is avoided (not supported by esbuild yet)
   - `queryWithParameters.mjs` now uses async iterables in a top-level `for-await` loop, which is possible when executing directly with node.js
   - `invokableScripts.mjs` uses async iterable, it is thus a bit simpler
   - a few other cosmetic changes in other examples
      - `follow-redirects.mjs` now demosntrate a `rowMapper` function
      - `index.html` is indented and produces better console logs
- typescript transpilation changed to produce EcmaScript2018 output, it is required to support async iterable and `for-await` loops, it was es2015 before this PR ... 
   - es2018 is properly implemented these days in modern browsers, deno and node.js environments
   - the situation is explained in the release notes
   - CI also shifted to node v16 (lts version until 2022-10-18)

## Checklist

- [x] CHANGELOG.md updated
- [x] Rebased/mergeable
- [x] A test has been added if appropriate
- [x] `yarn test` completes successfully
- [x] Commit messages are [conventional](https://www.conventionalcommits.org/en/v1.0.0/)

### CODE REVIEW

File: examples/query.ts [hunk 1]
```diff
@@ -10,16 +11,18 @@ const queryApi = new InfluxDB({url, token}).getQueryApi(org)
...
-// the essential ones are shown/commented below. See also rxjs-query.ts .
-//
-// Execute query and receive table metadata and rows as they arrive from the server.
-// https://docs.influxdata.com/influxdb/latest/reference/syntax/annotated-csv/
-queryApi.queryRows(fluxQuery, {
-  next: (row: string[], tableMeta: FluxTableMetaData) => {
+// the essential ones are shown in functions below.
+// Execution of a particular function follows
+// its defintion, comment/uncomment it at will.
+// See also rxjs-query.ts and queryWithParamas.mjs .
```
Comment (vlastahajek) [inline examples/query.ts:18]: I don't see `rxjs-query.ts` file

Comment (sranka) [reply examples/query.ts:18]: it is in the same directory

Comment (sranka) [reply examples/query.ts]: ... but there was a type in `queryWithParams.mjs` example
```bash
influxdb-client-js/examples$ ls rxjs-query.ts queryWithParams.mjs
queryWithParams.mjs  rxjs-query.ts
```

Comment (vlastahajek) [reply examples/query.ts]: > it is in the same directory

Ok, I thought `rxjs-query.ts` is a new file.

File: examples/invokableScripts.mjs [hunk 2]
```diff
@@ -69,26 +69,15 @@ async function invokeScript(scriptID) {
   // Use FluxScriptInvocationAPI to execute a particular
   // script with specified parametes and process parsed results
```
Comment (vlastahajek) [inline examples/invokableScripts.mjs:70]:
```suggestion
  // script with specified parameters and process parsed results
```

File: packages/core/src/results/chunksToLinesIterable.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import {ChunkCombiner, createTextDecoderCombiner} from './chunkCombiner'
+
+/**
+ * ChunksToLinesIterable is a transformation that accepts
+ * an iterable of Uint8Array instancesand returns iterable of lines.
```
Comment (vlastahajek) [inline packages/core/src/results/chunksToLinesIterable.ts:5]:
```suggestion
 * an iterable of Uint8Array instances and returns iterable of lines.
```

**Review by vlastahajek — APPROVED** [review]
Awesome! :clap:

File: examples/README.md [hunk 4]
```diff
@@ -4,15 +4,15 @@ This directory contains javascript and typescript examples for node.js, browser,
...
+    - [node](https://nodejs.org/en/) installed, at least version 16 is recommended
     - Run `npm install` in this directory
     - Change variables in [./env.mjs](env.mjs) to configure connection to your InfluxDB instance. The file can be used as-is against a new [docker InfluxDB v2.3 OSS GA installation](https://docs.influxdata.com/influxdb/v2.3/get-started/)
   - Examples are executable. If it does not work for you, run `npm run esr EXAMPLE.ts`.
   - [write.mjs](./write.mjs)
     Write data points to InfluxDB.
   - [query.ts](./query.ts)
     Query InfluxDB with [Flux](https://docs.influxdata.com/influxdb/latest/get-started/).
-  - [queryWithParams.ts](./queryWithParams.ts)
+  - [queryWithParams.mjs](./queryWithParams.mjs)
```
Comment (bednar) [inline examples/README.md:15]: Please also change the `[scripts/server.js](./scripts/server.js)` to `[scripts/server.mjs](./scripts/server.mjs)`.

**Review by bednar — CHANGES_REQUESTED** [review]
Thanks for your PR 👍 There is only minor requirement:

File: examples/README.md (hunk 4 above)
Comment (sranka) [reply examples/README.md:15]: nice catch :+1:

**Review by bednar — APPROVED** [review]
LGTM 🍻

**Review by bednar — APPROVED** [review]

**Comment by AzuladoToujours** [comment]
Nice! Just what i was searching for. Thanks.

**Pull request merged**